Validate and document 0.12 mbits/network deprecations

nickethier opened the pull request:
This PR adds two warnings to job validation including a check for task network resources and mbits in the group network. 

In addition a notice is added to the upgrade docs about the deprecation.

fixes #8497 #8681 #8746 #8780

arianvp commented:
Note that the homepage still uses `resources.network` as an example: https://www.nomadproject.io/docs/job-specification  I'm not sure if this is already fixed in master; but I found it odd as it uses a feature that isn't documented on the rest of the website anymore

nickethier commented:
I changed the mock Job to reflect the task network changes and ended up playing quite a bit of whack-a-test as a result. I also updated the init command to render jobs with network and service definitions in the task group with links back to docs site.

schmichael commented on command/assets/example-short.nomad:
```diff
@@ -2,6 +2,10 @@ job "example" {
   datacenters = ["dc1"]
 
   group "cache" {
+    network {
+      port "db" {}
+    }
+
```
When running locally against a dev agent (either root or non-root) I get:

```
$ nomad run https://raw.githubusercontent.com/hashicorp/nomad/6875248c1934185e9cbbe1485dcde6e69db280e5/command/assets/example-short.nomad
Job Warnings:
1 warning(s):

* Group "cache" has warnings: 1 error occurred:
        * mbits has been deprecated as of Nomad 0.12.0. Please remove mbits from the network block



==> Monitoring evaluation "0acf50e0"
...
```

The task fails to run with:

````
Trying to map ports but no network interface is available
```

nickethier replied:
:facepalm: I didn't update the examples with the new syntax. Fixed + added a more helpful error when `port_map` is used where `ports` should be

notnoop commented on nomad/structs/structs.go:
```diff
@@ -6581,12 +6570,10 @@ func validateServices(t *Task) error {
+		for portLabel := range ports {
+			portLabels[portLabel] = struct{}{}
 		}
 	}
```
Is there a backward compatibility concern here?  If an old job that specifies ports and labels in task resources, will they continue to be validated/processed the same way?

notnoop replied:
Checking here again - if the answer is no, let's merge then.

shoenig replied:
@notnoop I've re-added the removed block, appending task level port labels to the set of port labels to validate against.

notnoop commented on drivers/docker/driver.go:
```diff
@@ -1078,6 +1078,9 @@ func (d *Driver) createContainerConfig(task *drivers.TaskConfig, driverConfig *T
 	default:
 		if len(driverConfig.PortMap) > 0 {
+			if task.Resources.Ports != nil {
+				return c, fmt.Errorf("'port_map' cannot map group network ports, use 'ports' instead")
```
How will this clause be triggered?  It seems it's captured in the first switch case statement in line 1055 above:

```go
	case task.Resources.Ports != nil && len(driverConfig.Ports) > 0:
```

nickethier replied:
this clause is triggered if `port_map { ... }` is defined in the driver config (driverConfig.**PortMap**) and group network ports are defined (task.Resources.Ports)

The first case statement is hit if group network ports are defined and `ports = [...]` is defined in the driver config (driverconfig.**Ports**).

Its meant to catch the following configuration:

```hcl
group {
  network {
    port "http" { }
  }
  task {
    driver = "docker"
    config {
      port_map {
        # CANNOT reference group ports here, must use ports = ["http"] and set port.to = 8080
        http = 8080
      }
    }
  }
}
```

notnoop replied:
Makes sense - thanks!

notnoop reviewed (commented):
Overall looks good - a couple of questions nitpicking about versions.

nickethier commented on website/pages/docs/upgrade/upgrade-specific.mdx:
```diff
@@ -17,6 +17,20 @@ standard upgrade flow.
+### `mbits` and Task Network Resource deprecation
+
+Starting in Nomad 0.12.0 the `mbits` field of the network resource block has
+been deprecated and is no longer considered when making scheduling decisions.
```
In 0.12 it was removed from the ranker. So it didn't have any influence on a node's score.

shoenig commented:
I'm pretty sure this test failure is relevant, I'll dig into it. 

```
    --- FAIL: TestHTTP_JobPlanRegion/api_region_takes_precedence (0.14s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2672958]

goroutine 39139 [running]:
testing.tRunner.func1.1(0x2c52e20, 0x4dcf3f0)
	/usr/local/go/src/testing/testing.go:1057 +0x30d
testing.tRunner.func1(0xc008172600)
	/usr/local/go/src/testing/testing.go:1060 +0x41a
panic(0x2c52e20, 0x4dcf3f0)
	/usr/local/go/src/runtime/panic.go:969 +0x175
github.com/hashicorp/nomad/command/agent.ApiNetworkResourceToStructs(0xc00361c1a0, 0x1, 0x4, 0x0, 0x0, 0x0)
	/home/circleci/go/src/github.com/hashicorp/nomad/command/agent/job_endpoint.go:1196 +0x138
```

shoenig commented:
The updated documented examples seem to work as expected:

```
[nzxt foo] $ nomad job init
Example job file written to example.nomad
[nzxt foo] $ nomad job run example.nomad
==> Monitoring evaluation "80e90cfc"
    Evaluation triggered by job "example"
    Allocation "2f608bc0" created: node "d079b198", group "cache"
    Evaluation within deployment: "1182e4ad"
    Evaluation status changed: "pending" -> "complete"
==> Evaluation "80e90cfc" finished with status "complete"
```

```
[nzxt foo] $ nomad job init -short
Example job file written to example.nomad
[nzxt foo] $ nomad job run example.nomad
==> Monitoring evaluation "fe683a47"
    Evaluation triggered by job "example"
    Allocation "eae73160" created: node "d079b198", group "cache"
    Evaluation within deployment: "5d30c927"
    Allocation "eae73160" status changed: "pending" -> "running" (Tasks are running)
    Evaluation status changed: "pending" -> "complete"
==> Evaluation "fe683a47" finished with status "complete"

```

A warning when setting `task.resources.networks.mbits`:
```
task.resources.networks.mbits set
$ nomad job run example.nomad
Job Warnings:
1 warning(s):

* Group "cache" has warnings: 1 error occurred:
	* 1 error occurred:
	* Task "redis": task network resources have been deprecated as of Nomad 0.12.0. Please configure networking via group network block.
```

A warning when setting `group.networks.mbits`:
```
$ nomad job run example.nomad
Job Warnings:
1 warning(s):

* Group "cache" has warnings: 1 error occurred:
	* mbits has been deprecated as of Nomad 0.12.0. Please remove mbits from the network block
```

shoenig commented:
The port validation in `validateServices` that @mahmood pointed out might still be a problem. 

There's this difference in behavior between master and this PR using this job file:

```hcl
job "example" {
  datacenters = ["dc1"]

  group "cache" {
    task "redis" {

      service {
	name = "rservice"
	port = "db"
      }
      
      driver = "docker"

      config {
        image = "redis:3.2"

        port_map {
          db = 6379
        }
      }

      resources {
        cpu    = 500
        memory = 256

        network {
          mbits = 10
          port "db" {}
        }
      }
    }
  }
}
```

on master)
```
$ nomad job run example.nomad
==> Monitoring evaluation "a9dd7ade"
    Evaluation triggered by job "example"
    Allocation "695c295f" created: node "d123aad0", group "cache"
    Evaluation within deployment: "857f2f8f"
    Allocation "695c295f" status changed: "pending" -> "running" (Tasks are running)
    Evaluation status changed: "pending" -> "complete"
==> Evaluation "a9dd7ade" finished with status "complete"
```

on PR)
```
$ nomad job run example.nomad
Error submitting job: Unexpected response code: 500 (1 error occurred:
	* Task group cache validation failed: 1 error occurred:
	* Task redis validation failed: 1 error occurred:
	* 1 error occurred:
	* port label "db" referenced by services rservice does not exist
```

shoenig commented:
Added a fix for the above backwards incompatibility, now the job is submissible with warnings as expected

```
$ nomad job run example.nomad 
Job Warnings:
1 warning(s):

* Group "cache" has warnings: 1 error occurred:
	* 1 error occurred:
	* Task "redis": task network resources have been deprecated as of Nomad 0.12.0. Please configure networking via group network block.





==> Monitoring evaluation "6e8cc203"
    Evaluation triggered by job "example"
    Allocation "b3a2ccbe" created: node "1b39f488", group "cache"
==> Monitoring evaluation "6e8cc203"
    Evaluation within deployment: "c8339b4c"
    Allocation "b3a2ccbe" status changed: "pending" -> "running" (Tasks are running)
    Evaluation status changed: "pending" -> "complete"
==> Evaluation "6e8cc203" finished with status "complete"
```

shoenig commented:
I'll add a few e2e tests around this, but in a separate PR.

kpweiler commented:
@shoenig - I think the above workaround might be misunderstanding the problem. The problem is that services can be defined at the group OR the task level, but network resources defined at the task level are deprecated. If a port is defined at the group level, it can't seem to be used at the task level - hence:

For some context - I want to define my service at the task level in a "count > 1" situation so that I can apply a consul service tag to differentiate the different instances (using NOMAD_ALLOC_INDEX) - this yields DNS entries like `instance-0.myservice.service.consul` from `instance-${NOMAD_ALLOC_INDEX}.myservice.service.consul`. Maybe there's a better way to do this?
